Add support for video profiles

jehumb-havok opened the pull request:
Add the ability to select the video profile of a video capture device by unique string identifier, and to select a particular resolution and/or framerate within that profile.

For consistency, also add the ability to select the resolution and/or framerate when not using video profiles, as it seems that WebRTC media constraints are deprecated and being removed : https://bugs.chromium.org/p/webrtc/issues/detail?id=9239

jehumb-havok committed:
Add video profile selection in VideoCapturer

jehumb-havok committed:
Add video profile ID selection

Add an extra string argument to select a video profile by its unique
identifier, in addition of the constraints on width/height/framerate.

jehumb-havok committed:
Fix video profile ID in VideoCapturer init

jehumb-havok committed:
Add handling for other if block

jehumb-havok committed:
Only use video profile if requested

jehumb-havok committed:
Clean-up code, remove duplicates

jehumb-havok committed:
Clean-up unused captures in lambda

jehumb-havok committed:
Avoid duplicate MediaCapture creation

Prevent StartCapture() from shutting down and recreating the
MediaCapture object, which is always already initialized when entering
the method, and was resulting in unnecessary work and possible
discrepancies when recreating.

jehumb-havok committed:
Select video profile by ID only

Do not apply the resolution constraints when searching for a video
profile by unique ID, as the video profile ID is unique anyway so this
is unnecessary work.

Instead, apply constraints while listing supported resolutions, such
that only the resolutions matching the current constraints are returned
and can be selected by the internal WebRTC implementation.

TODO: Note that currently this returns resolutions not in the video profile.

jehumb-havok committed:
Restrict video formats to selected profile

Only list supported video formats which are associated with the
currently selected video profile.

djee-ms commented:
Ping @tadam-msft - you're the only alias I know, sorry.

tadam-msft commented:
Sorry in training all day tidy, I'll try to take a look.
________________________________
From: Jerome Humbert <notifications@github.com>
Sent: Wednesday, August 14, 2019 5:56:16 AM
To: webrtc-uwp/webrtc-apis <webrtc-apis@noreply.github.com>
Cc: Terry Adams <Terry.Adams@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [webrtc-uwp/webrtc-apis] Add support for video profiles (#24)


Ping @tadam-msft<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftadam-msft&data=02%7C01%7CTerry.Adams%40microsoft.com%7C88dd4e74a6424bac51d008d720b6cbe1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637013841780191044&sdata=h8yYSOSnBjtm6SmPbgy12qQyuUspHYQXThpcR7tA4kk%3D&reserved=0> - you're the only alias I know, sorry.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fwebrtc-uwp%2Fwebrtc-apis%2Fpull%2F24%3Femail_source%3Dnotifications%26email_token%3DAFC4PEGLV5RGZKZ4YX4WUSDQEP6HBA5CNFSM4ILURCS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4IWOQY%23issuecomment-521234243&data=02%7C01%7CTerry.Adams%40microsoft.com%7C88dd4e74a6424bac51d008d720b6cbe1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637013841780191044&sdata=5UqIB7p8gJZIO5HZnhdT74VgreqjkDkHl7JUnIWNiBw%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFC4PECWZYLRDALYLWD4MUDQEP6HBANCNFSM4ILURCSQ&data=02%7C01%7CTerry.Adams%40microsoft.com%7C88dd4e74a6424bac51d008d720b6cbe1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637013841780201038&sdata=d3FnZ7i5N7usSL5Xq%2BYVJFzAr1N5V5RaIxTHGakiMy8%3D&reserved=0>.

robin-raymond reviewed (approved):
Obviously this needs to be tested thoroughly first but the code changes are good from a visual perspective. @morosev can you review too please since this is most where you've been coding?

jehumb-havok committed:
Simplify video format gathering

Gather supported WebRTC video formats using
GetAvailableMediaStreamProperties(), which will return a filtered list
of media types once the MediaCapture object has been initialized with a
specific video profile.

jehumb-havok committed:
Ensure consistency between WebRTC and MediaCapture

Ensure that the media type selected by the driver from the MediaCapture
settings is always the same as the video format selected by WebRTC.

The potential discrepancy arises from the fact that setting the
RecordMediaDescription field on the MediaCapture init settins tells the
driver to use that media type, but WebRTC doesn't know anything about
thsi choice and would use a different heuristic to select a video
format.

To prevent that, if a video profile is selected, only return a single
supported video format corresponding to the one selected in the
MediaCapture object.

djee-ms commented:
Updated the code to use [GetAvailableMediaStreamProperties()](https://docs.microsoft.com/en-us/uwp/api/windows.media.devices.videodevicecontroller.getavailablemediastreamproperties) for supported video formats, and enforce consistency between WebRTC and MediaCapture.

First, when a MediaCapture object is initialized with a video profile, then GetAvailableMediaStreamProperties() will filter out the media types (= video formats) returned to those compatible with the selected profile. This makes the code much simpler and naturally consistent. This is also the recommended way according to the Media Foundation team.

However, WebRTC and MediaCapture have different ways of selecting a media type / video format to use:
- For MediaCapture, setting the [RecordMediaDescription](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.mediacaptureinitializationsettings.recordmediadescription) field on the init settings instructs the driver to use that media type by default. But WebRTC doesn't know about this.
- For WebRTC, the video format is selected with a heuristic involving the list of supported video formats returned by the device, and a set of media constraints.

There are 2 issues with this:
1. WebRTC and MediaCapture are not aware of each other's choice, and it can happen that MediaCapture opens the video device with a different resolution than the one expected by WebRTC, which will result in video corruption or crashes.
2. The media constraints API is deprecated, and removed in later versions of WebRTC, so is not the way to go.

As a result, the proposed change will only return to WebRTC a single supported video format corresponding to the media type selected in MediaCapture, which locks WebRTC into the only good choice of video format and prevents any possible discrepancy.

robin-raymond reviewed (commented):
Will do a final review when changes are ready.

morosev reviewed (approved):
Everything looks good to me. The selection of video properties is definitely better than it was before. We might use the same engine in M75, since the video deice format selection has been changed there as well.

jehumb-havok committed:
Add support for video profile kind

Allow the caller to specify a kind of video profile (a.k.a. well-known
profile) instead of an exact video profile ID. The former is easier to
use when some UI needs to display some choice. It maps to the UWP API
enum MediaCapture.KnownVideoProfile.

djee-ms commented:
Added support for [`KnownVideoProfile`](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.knownvideoprofile) which is nicer than GUID-based profile IDs when the app displays some UI, and is also device-independent.

Tested on HoloLens 2 (with profiles) and locally on PC with a USB webcam (no profile). This is ready for a final review @robin-raymond. Ping also @tadam-msft if you want to have a look.

jehumb-havok committed:
Allow using only video profile kind

Avoid throwing an error if the video profile is not found although it
was not specified, and only a known one was.

jehumb-havok committed:
Fix video profile kind selection logic

Allow selecting a video profile kind without an explicit video profile
ID, and filter profiles from the known ones based on the resolution and
framerate constraints.

djee-ms commented:
Ping @robin-raymond and @tadam-msft, can you please have a look before we merge? We have customers waiting for those changes.

robin-raymond commented:
@djee-ms Ok, I looked preliminary a while ago but I'll relook again today. Sorry I was away for a few days on holiday.

morosev commented on idl/VideoCapturer.idl:
```diff
@@ -93,13 +93,66 @@ namespace org
       int height;
     };
 
+	enum VideoProfileKind
```
Tab is used instead of spaces

morosev commented on windows/wrapper/impl_webrtc_IVideoCapturer.h:
```diff
@@ -61,7 +61,12 @@ namespace webrtc
       const char *name_ {};
       const char *id_ {};
+      const char *videoProfileId_ {};
+	    int videoProfileKind_ {};
```
Tab symbol

morosev commented on windows/wrapper/impl_webrtc_VideoCapturer.cpp:
```diff
@@ -838,91 +973,257 @@ namespace webrtc
+    if (result->init(info)) {
+	  return result;
+    }
+	return {};
```
Tab symbol

morosev commented on windows/wrapper/impl_webrtc_VideoCapturer.cpp:
```diff
@@ -838,91 +973,257 @@ namespace webrtc
+             "framerate";
+      device_ = nullptr;
+      return false;
+	}
```
Tab symbol

morosev commented on windows/wrapper/impl_webrtc_VideoCapturer.cpp:
```diff
@@ -838,91 +973,257 @@ namespace webrtc
+      return false;
+	}
     SetSupportedFormats(formats);
+	return true;
```
Tab symbol

morosev reviewed (approved):
I put just few comments related to formatting - using tab symbol instead of spaces. All other looks good to me.

jehumb-havok committed:
Replace tabs with spaces

robin-raymond commented:
The changes overall look good. I think this is likely to break dot net implementation. I'd have to clone it down, merge in, and compile the dot net to see what happens.

djee-ms commented:
@robin-raymond why do you think that? Video profiles are optional as implemented, so existing code should work. We are discussing with @jamescadd going ahead with the merge due to time constraints, so if .NET imlementation is going to break this will be a blocker.

robin-raymond commented:
Go ahead and merge. I will fix releases/m71 compilation issues should any occur. I don't expect any runtime issue at all.
